RedHatInsights/rhsm-subscriptions
# SWATCH-901, SWATCH-90: Introduce host_id column to instance tables

* Jira issue: [SWATCH-901](https://issues.redhat.com/browse/SWATCH-901)
* Jira issue: [SWATCH-90](https://issues.redhat.com/browse/SWATCH-90)

Description
===========

Because it's currently confusing to have a `instance_id` column on `instance_measurements` and `instance_monthly_totals` that refers to `host.id` rather than `host.instance_id`.

This change adds renames the column and then adds a computed `instance_id` column in order to keep the schema compatible with the previous deployment.

Previous deployment writes are expected to fail, but reads (API) will succeed.

Note: I did some quick performance testing and determined that with 10,000,000 `instance_measurements` records (within the order of magnitude of prod data), it'll take around 25 seconds to complete, blocking the table for that long.

If curious, you can use the below (along with the host record from testing steps) to do similar:

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
INSERT INTO
  instance_measurements (instance_id, uom, value)
SELECT
  '78ef3f29-c900-4b9f-af04-e9aeb0231daf', i::varchar, i
FROM
  generate_series(1,10000000) i;
EOF
```

`instance_monthly_totals` in prod is small enough to not be a concern.

There is a follow-up changeset to be uncommented and run for a subsequent deployment.

Testing
=======

Setup
-----

1. Checkout `main`
2. Ensure that the DB schema is in place:

```shell
./gradlew :liquibaseUpdate
```

3. Insert some test data:

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
insert into account_config(org_id, opt_in_type, created, updated) values ('org123', 'JMX', now(), now());
insert into account_services(org_id, service_type) values ('org123', 'HBI_HOST');
insert into hosts(id, inventory_id, account_number, insights_id, org_id, display_name, subscription_manager_id,
                  is_guest, hypervisor_uuid, hardware_type, num_of_guests, last_seen, is_unmapped_guest, is_hypervisor,
                  cloud_provider, instance_id, billing_provider, billing_account_id)
values ('78ef3f29-c900-4b9f-af04-e9aeb0231daf', 'd25b4de0-965a-462e-8f0e-6f51d1fd9590', 'account123', '9efad409-9366-48b1-8079-e33333a4715e', 'org123', 'example', '75dabfd8-c67c-45a2-bb19-9c11bb68c242',
        false, null, 'PHYSICAL', null, now(), false, false,
        null, '44d3e412-b238-4de3-ab02-e4b87ea00466', null, null);
insert into instance_measurements (instance_id, uom, value)
  values ('78ef3f29-c900-4b9f-af04-e9aeb0231daf', 'CORES', 41.5);
insert into instance_monthly_totals(instance_id, month, uom, value)
  values ('78ef3f29-c900-4b9f-af04-e9aeb0231daf', '2023-07', 'CORES', 42.5);
insert into host_tally_buckets(host_id, product_id, sla, as_hypervisor, measurement_type)
  values ('78ef3f29-c900-4b9f-af04-e9aeb0231daf', 'OpenShift-metrics', '_ANY', false, 'PHYSICAL');
insert into host_tally_buckets(host_id, product_id, sla, as_hypervisor, measurement_type)
  values ('78ef3f29-c900-4b9f-af04-e9aeb0231daf', 'OpenShift Container Platform', '_ANY', false, 'PHYSICAL');
EOF
```

4. Insert some hosts into HBI for testing nightly tally:

```shell
bin/insert-mock-hosts \
  --num-hypervisors=1 \
  --num-guests=1 \
  --hbi \
  --org org123 \
  --clear
```

5. Insert some events for testing hourly tally later:

```shell
bin/import-events.py --file bin/events.csv
```

Testing - API
-------------

1. Checkout this branch.
2. Run DB migrations:
```shell
./gradlew :liquibaseUpdate
```
3. Checkout `main`.
4. Deploy:

```shell
DEV_MODE=true ./gradlew :bootRun
```

5. Try the instances API for OpenShift-metrics (returning monthly totals):

```shell
http :8000/api/rhsm-subscriptions/v1/instances/products/OpenShift-metrics \
  beginning==2023-07-01T00:00Z \
  ending==2023-07-31T23:59:59.99999Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id": "org123"}}}' | base64 -w0)
```

Observe 42.5 in the response.

6. Try the instances API for OpenShift Container Platform (returning instance_measurements):

```shell
http ":8000/api/rhsm-subscriptions/v1/instances/products/OpenShift Container Platform" \
  uom==CORES \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id": "org123"}}}' | base64 -w0)
```

Observe 41.5 in the response.

7. Repeat steps 4-6 for this branch.

Testing - nightly tally works
-----------------------------

1. From this branch, initiate nightly tally:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/internal/rpc/tally/snapshots/org123 \
  x-rh-swatch-psk:placeholder
```

2. Verify that the tally completed successfully (check in a few seconds to give tally a chance to finish, you'll see some instance_measurements recorded):

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
select * from instance_measurements;
EOF
```

Testing - hourly tally works
----------------------------

1. From this branch, initiate hourly tally for the mock events:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/internal/rpc/tally/snapshots \
  begin==2021-10-18T00:00:00Z \
  end==2021-10-19T00:00:00Z \
  x-rh-swatch-psk:placeholder
```

2. Verify that the tally completed successfully (check in a few seconds to give tally a chance to finish, you'll see some instance_montly_totals recorded):

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
select * from instance_monthly_totals;
EOF
```

Cleanup
-------

Do a liquibase rollback in order to revert the migrations:

```shell
./gradlew :liquibaseRollbackCount -PliquibaseCommandValue=2
```
